### ISSUE TITLE
How to support other projection?

### ISSUE THREAD

**Issue by KingElimy** [mapbox/tippecanoe#546]
Hi eric,
how can i simply support some projections,such as cgcs2000, or is there planning to add projection transform both input and output?
And I'm confused how the function [epsg3857totile](https://github.com/mapbox/tippecanoe/blob/3cea1257a5704535c351eb93783d6c9d9422440f/projection.cpp#L51) works, can you leave me some tips?
thanks sincerely!

**Comment by ericfischer** [comment]
I have no current plans to add more projections. Are you wanting to convert CGCS2000 data to the standard Web Mercator of vector tiles, or are you wanting the data to remain in that projection even after tiling?

If you want non-Mercator tiles, https://github.com/mapbox/tippecanoe/issues/422 has some examples of using other projections denominated in meters. If you want Mercator tiles, I would suggest using `ogr2ogr` to convert your file's original projection to WGS84.

The `epsg3857totile` function is scaling and offsetting coordinates from their original unit (meters) into Tippecanoe's internal units (1/(2^32) of the circumference of the earth). The constant 6378137.0 is the radius of the Earth in meters.

Basically:

```
coordinate (meters)
* 1 radius / 6378137.0 meters
* 1 diameter / 2 radius
* 1 circumference / pi diameter
* 2^32 tiles / circumference
```

except that it uses 2^31 instead of the two separate operations of multiplying by 2^32 and dividing by 2.

The extra subtraction for the Y axis is because the axis runs in the opposite direction in tile coordinates (north pole is 0) as it does in EPSG 3857 coordinates (equator is 0).

**Comment by KingElimy** [comment]
oh,I should have a try what you've illustrated in #422.
what i want is non-Mercator, since I search the key word of "projection" and there is no list of #422 ,I get the source code modified where lonlat2tile[(https://github.com/mapbox/tippecanoe/blob/6b29966e4ae339881c795d5cd9a9d45d9c34fe6e/projection.cpp#L38)]by instead y to 
`long long lly = (n>>1) * ((-lat + 90) /180);` 
and function tile2lonlat does the opposite, Is that the right way to handle it?

**Comment by ericfischer** [comment]
That sounds like it would make plate carrée tiles, so yes, if that's what you're trying to do.

For my own plate carrée experiment, as it says in #422, instead of modifying the code I reprojected the input:

```
$ ogr2ogr -f GeoJSON -t_srs EPSG:32663 pc.json ne_10m_admin_0_countries.json
$ tippecanoe -f -o foo.mbtiles -X -z5 -s EPSG:3857 pc.json
```

**Comment by KingElimy** [comment]
Much help,thanks :)

**Comment by mjj203** [comment]
what if you add an additional function called lonlat2tile3395 so that DoD customers can more easily convert from their raw WGS84/EPSG4326 data into EPSG3395 mbtiles? something like:
```
void lonlat2tile3395(double lon, double lat, int zoom, long long *x, long long *y) {
    int a = 6378137; // WGS84 semi-major axis
    double b = 6356752.3142; // WGS84 semi-minor axis
    long long e = sqrt(1 - b^2 / a^2); // ellipsoid eccentricity
    long long c = pow((1 - e*sin(lat)) / (1 + e*sin(lat)), e/2);
    unsigned long long n = 1LL << zoom;
	long long llx = n * (a * lon);
	long long lly = n * (a * ln(tan(M_PI_4 + lat/2) * c));

	*x = llx;
	*y = lly;
}
```
Seems like any projected coordinate system can be easily hacked by pre-projecting the geojson, but what about creating un-projected mbtiles to keep them in the raw WGS84 DD format?

The DoD legally can't use EPSG3857 due to it being almost 40,000m off in some places. [NGA Warning](http://earth-info.nga.mil/GandG/wgs84/web_mercator/index.html) NGA has a c++ coordinate conversion tool called Geotrans that is the officially authorized conversion tool. [geotrans](http://earth-info.nga.mil/GandG/geotrans/)

**Comment by KingElimy** [comment]
In my opinion,just the tile make process,it's ok. But, you should consider that when you show the tile data in any canvas,and then to locate where (mean,the real location you pointer to). If you just show it and won't to overlay other vector data, then it may works well, isn't?
For your case,when use the tile data to reverse calculate the geographic coordinate from the projected coordinate,just make up the offset, say 40000 meters,  then it will be right.
Am i right? @ericfischer @mjj203

**Comment by mjj203** [comment]
That isn't the answer troops in war zones want to hear when they are building conops or calling in air support and need to decide if that hellfire should be 40,000m to the left or right. The point is vector tile data is the way forward in transmitting and rendering large amounts of geospatial data, and if its going to be used for things like self driving vehicles in the civilian market, then it needs to be navigationally safe. @KingElimy

**Comment by KingElimy** [comment]
It is important that the presentation and slicing need to work together on projection.
